### DISCUSSION TITLE
0.35.0 refactor of root_path handling is potentially returning incorrect route

### DISCUSSION THREAD

**Discussion by robbielaw** [encode/starlette#2495] [Potential Issue]
After we applied the FastAPI 0.109.0 update released on Jan 11th, we began seeing certain endpoints in our API return a 404. This issue didn't appear when running the API locally, only when accessing via the API gateway proxy (which we supply a `root_path` to our FastAPI instance).

Part of the FastAPI 0.109.0 update included the move to Starlette 0.35.0. The introduction of `get_route_path`  (#2400) in 0.35.0 attempts to strip the `root_path` from the beginning of the active route. However, if the active route also contains the string that is the `root_path` somewhere else within it, it results in it also being stripped out. 

**Example Scenario**:
`root_path` for the API is `/product`. A router(s) has the prefix `/products`. That router has an endpoint of `/detail`. This should produce the route of `product/products/detail`. I believe the intention of `get_route_path` is to then return the route `/products/detail`, but its current implementation will return `s/detail`. This results in, when the endpoint is called, returning a 404.

**Comment by Pentusha** [comment]
Faced with the same issue.
Described an issue in the PR #2400. Attached an example project with tests on which regression can be reproduced.

**Reply by jaspersival** [reply]
Same problem, working with a root_path of /api and trying to mount static files at /prefix/static/my_static_file should return the static file at /api/prefix/static/my_static_file. However, I get a 404 error. I am slamming my head on this for days already.

**Reply by smrtl** [reply]
We have the same issue. I wonder why Pentusha's comment in #2400 was ignored and the PR was merged nonetheless.

**Reply by DamJanCzarek** [reply]
Bump, we got the same problem in our project :(

**Reply by gabriel-f-santos** [reply]
We run in the same problem.

Its possible to fix this, making the regex look just for the root_path:

```
def get_route_path(scope: Scope) -> str:
    root_path = scope.get("root_path", "")
    route_path = re.sub(r"^" + root_path + r"(?=/|$)", "", scope["path"])
    return route_path
```

Just one parameter more concatenated in the regex. None test was broken with this.

**_+ r"(?=/|$)"_**

Cases:
scope = {"path": "/foo-123/bar", "root_path": "foo"}
print(get_route_path(scope)) # Output: '/foo-123/bar'

scope = {"path": "/foo/bar", "root_path": "foo"}
print(get_route_path(scope)) # Output: '/bar'

scope = {"path": "/foo", "root_path": "foo"}
print(get_route_path(scope)) # Output: ''
